File: android/src/main/kotlin/com/edfapg/flutter/sdk/eventhandlers/RecurringSaleEventHandler.kt
package com.edfapg.flutter.sdk.eventhandlers

import com.edfapg.flutter.sdk.helper.toMap
import com.edfapg.sdk.core.EdfaPgSdk
import com.edfapg.sdk.model.request.options.EdfaPgRecurringOptions
import com.edfapg.sdk.model.request.order.EdfaPgOrder
import com.edfapg.sdk.model.response.base.error.EdfaPgError
import com.edfapg.sdk.model.response.sale.*
import com.google.gson.Gson
import io.flutter.plugin.common.EventChannel


class RecurringSaleEventHandler: EventChannel.StreamHandler {
    var sink:EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        (arguments as? Map<*, *>)?.let {
            print(
                with(it) {
                    (get("EdfaPgOrder") as? Map<*, *>)?.let { orderMap ->
                        (get("payerEmail") as? String)?.let { payerEmail ->
                            (get("cardNumber") as? String)?.let { cardNumber ->
                                (get("EdfaPgRecurringOptions") as? Map<*, *>)?.let { recOptionMap ->
                                    (get("auth") as? Boolean)?.let { auth ->
                                        val order = Gson().fromJson(Gson().toJson(orderMap), EdfaPgOrder::class.java)
                                        val recurringOptions = Gson().fromJson(Gson().toJson(recOptionMap), EdfaPgRecurringOptions::class.java)

                                        recurringSale(auth = auth, order = order, payerEmail = payerEmail, cardNumber = cardNumber, recurringOptions = recurringOptions);

                                        "All params are valid"
                                    } ?: "Missing 'Boolean' auth parameter"
                                } ?: "Missing 'EdfaPgRecurringOptions' auth parameter"
                            } ?: "Missing 'cardNumber' parameter"
                        } ?: "Missing 'payerEmail' parameter"
                    } ?: "Missing 'EdfaPgOrder' parameter"
                }
            )
        }

    }

    override fun onCancel(arguments: Any?) {
    }

    fun recurringSale(auth:Boolean, order:EdfaPgOrder, payerEmail:String, cardNumber:String, recurringOptions:EdfaPgRecurringOptions){
        EdfaPgSdk.Adapter.RECURRING_SALE.execute(
            auth = auth,
            order = order,
            cardNumber = cardNumber,
            options = recurringOptions,
            payerEmail = payerEmail,
            callback = object : EdfaPgSaleCallback {
                override fun onResponse(response: EdfaPgSaleResponse) {
                    send(mapOf("responseJSON" to response.toMap()))
                    super.onResponse(response)
                }

                override fun onResult(result: EdfaPgSaleResult) {
                    val res = result.result
                    when (res) {
                        is EdfaPgSaleDecline -> send(mapOf("decline" to res.toMap()))
                        is EdfaPgSaleRecurring -> send(mapOf("recurring" to res.toMap()))
                        is EdfaPgSaleRedirect -> send(mapOf("redirect" to res.toMap()))
                        is EdfaPgSale3Ds -> send(mapOf("secure3d" to res.toMap()))
                        is EdfaPgSaleSuccess -> send(mapOf("success" to res.toMap()))
                    }
                }

                override fun onError(error: EdfaPgError){
                    send(mapOf("error" to error.toMap()))
                }

                override fun onFailure(throwable: Throwable) {
                    send(mapOf("failure" to throwable.toMap()))
                    super.onFailure(throwable)
                }
            }
        )

    }
    private fun send(map:Map<*,*>){
        sink?.success(map)
    }
}

File: ios/Classes/EventHandlers/RecurringSaleEventHandler.swift
//
//  SaleEventHandler.swift
//  edfapay_sdk
//
//  Created by Zohaib Kambrani on 03/03/2023.
//

import Foundation
import Flutter
import UIKit
import EdfaPgSdk


class RecurringSaleEventHandler : NSObject, FlutterStreamHandler{
    
    var eventSink:FlutterEventSink? = nil
    
    private lazy var recurringAdapter: EdfaPgRecurringSaleAdapter = {
        let adapter = EdfaPgAdapterFactory().createRecurringSale()
        adapter.delegate = self
        return adapter
    }()
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        
        if let params = arguments as? [String:Any],
           let auth = params["auth"] as? Bool,
           let payerEmail = params["payerEmail"] as? String,
           let cardNumber = params["cardNumber"] as? String,
           let order = params["EdfaPgOrder"] as? [String : Any?],
           let recurringOptions =  params["EdfaPgRecurringOptions"] as? [String : Any?]{
                 
            recurringAdapter.delegate = self
            recurringAdapter.execute(
                order: EdfaPgOrder.from_(dictionary: order),
                options: EdfaPgRecurringOptions.from(dictionary: recurringOptions),
                payerEmail: payerEmail,
                cardNumber: cardNumber,
                auth: auth,
                callback: handleResponse
            )
            
        }
        
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
    
    private func handleResponse(response: EdfaPgResponse<EdfaPgSaleResult>){
        
        switch response {
        case .result(let result):
            
            switch result {
            case .recurring(let result):
                let json = result.toJSON(root: "recurring")
                eventSink?(json)
                
            case .secure3d(let result):
                let json = result.toJSON(root: "secure3d")
                eventSink?(json)
                
            case .redirect(let result):
                let json = result.toJSON(root: "redirect")
                eventSink?(json)
                
            case .success(let result):
                let json = result.toJSON(root: "success")
                eventSink?(json)
                
            case .decline(let result):
                let json = result.toJSON(root: "decline")
                eventSink?(json)

            default: break
                let json = ["failure" : ["error" : "Unhandled response case at EdfaPaySaleResult.result"]]
                eventSink?(json)
                
            }
                        
        case .error(let error):
            let json = ["error" : error.json()]
            eventSink?(json)
            print(error)
            
        case .failure(let exception):
            if let err = exception as? NSError{
                let json = ["failure" : err.userInfo]
                eventSink?(json)
            }else{
                let json = ["failure" : ["exception" : exception.localizedDescription]]
                eventSink?(json)
            }
            print(exception)
            
        default:
            let json = ["failure" : ["error" : "Unhandled response case at EdfaPayResponse.result"]]
            eventSink?(json)
        }
    }

}


extension RecurringSaleEventHandler : EdfaPgAdapterDelegate{
    
    func willSendRequest(_ request: EdfaPgDataRequest) {
        
    }
    
    func didReceiveResponse(_ reponse: EdfaPgDataResponse?) {
        if let data = reponse?.data,
           let dict = try? JSONSerialization.jsonObject(with: data, options: .fragmentsAllowed){
            eventSink?(["responseJSON" : dict])
        }
    }
    
}
